pygen/pygen_src/riscv_instr_stream.py

```diff
@@ -1,237 +1,229 @@
+        return s
+
+
+class riscv_rand_instr_stream(riscv_instr_stream):
+    """
+    Generate a random instruction stream based on the configuration
+    There are two ways to use this class to generate instruction stream
+        1. For short instruction stream, you can call randomize() directly.
+        2. For long instruction stream (>1K), randomize() all instructions together might take a
+           long time for the constraint solver. In this case, you can call gen_instr to generate
+           instructions one by one. The time only grows linearly with the instruction count
+    """
+
+    def __init__(self):
+        # calling super constructor
+        super().__init__()
+        self.kernel_mode = 0
+        self.allowed_instr = []
+        self.category_dist = []
+
+    def create_instr_instance(self):
+        for i in range(self.instr_cnt):
+            self.instr_list.append(None)
+
+    def setup_allowed_instr(self, no_branch=0, no_load_store=1):
+        self.allowed_instr = riscv_instr_ins.basic_instr
+        if no_branch == 0:
+            self.allowed_instr.extend(
+                riscv_instr_ins.instr_category[riscv_instr_category_t.BRANCH.name])
+        if no_load_store == 0:
+            self.allowed_instr.extend(
+                riscv_instr_ins.instr_category[riscv_instr_category_t.LOAD.name])
+            self.allowed_instr.extend(
+                riscv_instr_ins.instr_category[riscv_instr_category_t.STORE.name])
+        self.setup_instruction_dist(no_branch, no_load_store)
+
+    # TODO
+    def randomize_avail_regs(self):
+        pass
+
+    def setup_instruction_dist(self, no_branch=0, no_load_store=1):
+        if cfg.dist_control_mode:
+            self.category_dist = cfg.category_dist
+            if no_branch:
+                self.category_dist[riscv_instr_category_t.BRANCH.name] = 0
+            if no_load_store:
+                self.category_dist[riscv_instr_category_t.LOAD.name] = 0
+                self.category_dist[riscv_instr_category_t.STORE.name] = 0
+            logging.info("setup_instruction_dist: %0d", len(self.category_dist))
+
+    def gen_instr(self, no_branch=0, no_load_store=1, is_debug_program=0):
+        self.setup_allowed_instr(no_branch, no_load_store)
+        for i in range(len(self.instr_list)):
+            self.instr_list[i] = self.randomize_instr(self.instr_list[i], is_debug_program)
+        while self.instr_list[-1].category == riscv_instr_category_t.BRANCH:
+            self.instr_list.pop()
+            if len(self.instr_list):
+                break
+
+    def randomize_instr(self, instr, is_in_debug=0, disable_dist=0):
+        exclude_instr = []
+        is_SP_in_reserved_rd = riscv_reg_t.SP in self.reserved_rd
+        is_SP_in_reserved_regs = riscv_reg_t.SP in cfg.reserved_regs
+        is_SP_in_avail_regs = riscv_reg_t.SP in self.avail_regs
+        if ((is_SP_in_reserved_rd or is_SP_in_reserved_regs) or (not is_SP_in_avail_regs)):
```

Comment:
the SV generator has a fourth check `avail_regs.size() > 0` here also, is this needed?

Reply:
Yes, We missed that one